docs/enhancements/support-node-labeling.md

```diff
@@ -0,0 +1,159 @@
+---
+
+# Support node labeling during installation
+
+## Summary
+
+There was a requirement to support node roles during installation.  These roles 
+are extra roles, which are used by the user and not by any Openshift components.
+These labels should be applied on a node in a spoke cluster.
+
+Although the requirement is specifically for roles, the implementation should support setting 
+any label on a node in a spoke cluster, regardless if the label represents a role.
+
+## Motivation
+
+Currently, nodes can be assigned with a single role which can be one of two values [ **master**, **worker**].
+The may want to add additional node labels that can be used as custom roles.  
+
+Each role in OCP are associated with machine config pool (MCP).  When a node reboots itself
+from the pointer ignition, it uses specific MCP that contains the specific configuration
+for the role.  This MCP has the format `rendered-<role>-*` (* stands for a hash string).  So there is MCP for master 
+role `rendered-master-*` and for worker `rendered-worker-*`.
+
+When adding custom role, it can be associated with custom MCP which references specific configuration
+for that role.  Setting appropriate label a node, causes the node to reboot and apply the 
+MCP configuration (machine config) on that node.
+
+Setting node labels should support this flow.
+
+### Goals
+- For ZTP flow: add capability to add and update annotations on BMH which will be transferred 
+to agent CR.
+- For boot it yourself flow: add capability to add and update labels to agent CR after agent
+  has been created.
+- Transfer the labels from the agent CR to the corresponding node on the spoke cluster.
+
+### Non goals
+- Single way to set node labels before agent creation both for ZTP flow, and boot it yourself flow.
+
+## Proposal
+
+### ZTP flow
+Labels can be applied on every CR.  So there is no known limitation to add a label on agent.
+
+For the ZTP flow, there should a pattern for transferring an annotation from the BMH to a label  
+in the corresponding agent.  The pattern is actually the prefix of the annotation.
+It should be **node-label**.  The suffix of the annotation will be transferred 
+to labels on the corresponding agent CR.
+
+For example assuming we want to apply a label **node-role.kubernetes.io/infra** on an
+a node.  Then the following annotation should be added to the BMH: **node-label.node-role.kubernetes.io/infra**.
+
+### Node labels on agent CR
+Additional field **nodeLabels** will be added to the spec 
+section of the agent CR.  This field will be a string list.
+
+For the ZTP flow, the annotations having the prefix **node-label** in BMH will be transferred
+to the dedicated field **nodeLabels** without the **node-label** prefix.
+
+If boot it yourself flow is in use, nodeLabels should be set directly on the agent CR.
+
+### Moving labels to nodes
+
+#### Day 1 cluster
+When installing day 1 cluster, the implementation should make use of the **nodeLabels** field
```

Comment:
Why nodeLabels and not machineConfigPoolName?

Reply:
I assume both. User will need to set mcp + role in nodelables

Reply:
@ori-amizur , this enhancement needs more details.
- The title indicates that the end goal is labeling nodes, but that's different from what the use case is.
- It would help to have a description of the steps required to do this with UPI and then what the steps would be with CIM after this proposal.  Here is a partial list of questions in my head after reading this: Does the user need to create an MCP?  How?  Does the user also need to use the machineConfigPoolName API?  How?  What is that API for and what is this label for?  If we need both machineConfigPoolName and label, can the existing API just add the label?

Reply:
Node labels can be used for many purposes - not necessarily roles.  This enhancement is for node labels, which custom roles are specific use-case for node labels.

Reply:
@avishayt although in terms of user story, MCP and MC are related to node labels, creation of MC and MCP are done by creating manifests.  This is already supported by AI for day1 installation.
As I indicated in the document, we should also support it for day2

Reply:
MCP name is another field that already exists. No need to add it.